src/java/org/apache/cassandra/index/sai/disk/vector/CompactionGraph.java

```diff
@@ -250,6 +259,38 @@ public InsertionResult maybeAddVector(ByteBuffer term, int segmentRowId) throws
             int ordinal = useSyntheticOrdinals ? nextOrdinal++ : segmentRowId;
             postings = new CompactionVectorPostings(ordinal, segmentRowId);
             postingsMap.put(vector, postings);
+
+            // fine-tune the PQ if we've collected enough vectors
+            if (!pqFinetuned && postingsMap.size() >= PQ_TRAINING_SIZE)
+            {
+                final var trainingVectors = new ConcurrentVectorValues(dimension);
+                postingsMap.forEach((v, p) -> trainingVectors.add(p.getOrdinal(), v));
+                // lock the addGraphNode threads out
+                trainingLock.writeLock().lock();
+                try
+                {
+                    // Fine tune the pq codebook and re-encode the vectors added so far.
+                    compressor = compressor.refine(trainingVectors);
+                    pqVectors = new MutablePQVectors(compressor, postingsEntriesAllocated);
+                    for (int i = 0; i < builder.getGraph().getIdUpperBound(); i++)
+                    {
+                        var v = trainingVectors.getVector(i);
+                        if (v == null)
+                            pqVectors.setZero(i);
+                        else
+                            pqVectors.encodeAndSet(i, v);
```

Comment:
Do we want to consider parallelizing this task like we do elsewhere? I think it might be worth it since dedicated compactors can sometimes only have one compaction running at a time, which implies we'd have available cpu cycles.

Reply:
you're probably right, but the mechanics are dicey?

i can make just vectorCount atomic in MutablePQ, in which case it would work for us but no other multithreaded case, or redo the whole thing with atomics which is ugly and probably a negative for performance in the common case

i guess the thing to do is to add a multithreaded method to MutablePQ that does exactly what we want